File: React_App/crypto.js
/**
 * EchoCrypt crypto layer — browser WebCrypto AES-256-GCM
 *
 * Byte layout (must match Python crypto.py exactly):
 *   nonce(12) + ciphertext + tag(16)
 *
 * WebCrypto AES-GCM expects the tag appended to ciphertext in one blob —
 * so we pass data.slice(12) directly (includes both ciphertext and tag).
 */

const SALT = new TextEncoder().encode('acoustinet_v1')
const ITERATIONS = 100_000

/**
 * Derive a 256-bit AES-GCM key from a passphrase using PBKDF2-SHA256.
 * This is intentionally slow (100k iterations) — call once, cache the result.
 *
 * @param {string} password
 * @returns {Promise<CryptoKey>}
 */
export async function deriveKey(password) {
  const keyMaterial = await crypto.subtle.importKey(
    'raw',
    new TextEncoder().encode(password),
    'PBKDF2',
    false,
    ['deriveKey']
  )

  return crypto.subtle.deriveKey(
    {
      name: 'PBKDF2',
      salt: SALT,
      iterations: ITERATIONS,
      hash: 'SHA-256',
    },
    keyMaterial,
    { name: 'AES-GCM', length: 256 },
    false,           // not extractable
    ['encrypt', 'decrypt']
  )
}

/**
 * Decrypt AES-256-GCM data.
 *
 * @param {Uint8Array} data  nonce(12) + ciphertext + tag(16)
 * @param {CryptoKey}  key   result of deriveKey()
 * @returns {Promise<Uint8Array>}
 * @throws if authentication fails or data is malformed
 */
export async function decrypt(data, key) {
  const nonce = data.slice(0, 12)
  // WebCrypto expects ciphertext||tag as one blob starting at offset 12
  const ciphertextWithTag = data.slice(12)

  const plaintext = await crypto.subtle.decrypt(
    { name: 'AES-GCM', iv: nonce },
    key,
    ciphertextWithTag
  )

  return new Uint8Array(plaintext)
}

/**
 * Encrypt data with AES-256-GCM.
 * Output: nonce(12) + ciphertext + tag(16) — matches Python crypto.py exactly.
 *
 * @param {Uint8Array} data  plaintext bytes
 * @param {CryptoKey}  key   result of deriveKey()
 * @returns {Promise<Uint8Array>}
 */
export async function encrypt(data, key) {
  const nonce = crypto.getRandomValues(new Uint8Array(12))
  const ct    = await crypto.subtle.encrypt({ name: 'AES-GCM', iv: nonce }, key, data)
  const out   = new Uint8Array(12 + ct.byteLength)
  out.set(nonce, 0)
  out.set(new Uint8Array(ct), 12)
  return out
}

File: Python/crypto.py
"""
EchoCrypt crypto layer — AES-256-GCM + PBKDF2-SHA256
Cross-compatible with browser WebCrypto (see byte layout note below).

Byte layout: nonce(12) + ciphertext + tag(16)
  - WebCrypto AES-GCM expects: nonce separate, then ciphertext||tag as one blob
  - pycryptodome separates tag — we reassemble in the correct order here
"""

from Crypto.Protocol.KDF import PBKDF2
from Crypto.Hash import SHA256
from Crypto.Cipher import AES
import os


SALT = b'acoustinet_v1'
ITERATIONS = 100_000


def derive_key(password: str) -> bytes:
    """Derive a 256-bit key from a passphrase using PBKDF2-SHA256."""
    return PBKDF2(
        password.encode(),
        SALT,
        dkLen=32,
        count=ITERATIONS,
        hmac_hash_module=SHA256
    )


def encrypt(plaintext: bytes, key: bytes) -> bytes:
    """
    Encrypt plaintext with AES-256-GCM.
    Returns: nonce(12) + ciphertext + tag(16)
    This layout is compatible with WebCrypto's AES-GCM decrypt.
    """
    nonce = os.urandom(12)          # ← force 12 bytes
    cipher = AES.new(key, AES.MODE_GCM, nonce=nonce)
    ciphertext, tag = cipher.encrypt_and_digest(plaintext)
    return nonce + ciphertext + tag

def decrypt(data: bytes, key: bytes) -> bytes:
    """
    Decrypt AES-256-GCM data.
    Expects: nonce(12) + ciphertext + tag(16)
    Raises ValueError if authentication fails.
    """
    
    nonce = data[:12]
    ciphertext = data[12:-16]
    tag = data[-16:]
    cipher = AES.new(key, AES.MODE_GCM, nonce=nonce)
    return cipher.decrypt_and_verify(ciphertext, tag)
